File: solutions/python/robot-name/1/robot_name.py
import string
from random import choice


class Robot:
    def __init__(self):
        self.name = self._generate_name()
        self.used_name = [self.name]

    def name(self):
        return self.name

    def reset(self):
        self.name = self._generate_name()

        if self.name in self.used_name:
            self.reset()

        self.used_name.append(self.name)

        return self

    @staticmethod
    def _generate_name():
        name = [choice(string.ascii_uppercase) for _ in range(2)]
        name += [str(choice(range(10))) for _ in range(3)]

        return "".join(name)

File: solutions/ruby/robot-name/1/robot_name.rb
NAME_TAKEN = []

class Robot
  attr_reader :name

  def initialize
    generate_name
    puts 'test'
  end

  def reset
    NAME_TAKEN.delete @name
    generate_name
  end

  def self.forget
    NAME_TAKEN.clear
  end

  private

  def generate_name
    loop do
      letters = ('A'..'Z').to_a.sample(2).join
      numbers = rand(100..999).to_s
      @name = letters + numbers
      break unless NAME_TAKEN.include?(@name)
    end

    NAME_TAKEN << @name
  end
end
